x-pack/legacy/plugins/spaces/server/lib/saved_objects_client/get_namespace.ts

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsBaseOptions } from 'src/core/server';
+import { DEFAULT_SPACE_ID } from '../../../common/constants';
+import { DEFAULT_SPACE_NAMESPACE } from './default_space_namespace';
+
+export function getNamespace(
+  operationOptions: SavedObjectsBaseOptions,
+  currentSpaceId: string
+): string | undefined {
+  if (operationOptions.namespace) {
+    if (operationOptions.namespace === DEFAULT_SPACE_NAMESPACE) {
```

Comment:
Who writes `DEFAULT_SPACE_NAMESPACE` in namespace field? If it serializable a symbol cannot be used. Can we use a special string instead? What is the benefit of using a separate type?
`DEFAULT_SPACE_ID` exists here only for backward compatibility?

Reply:
> Who writes DEFAULT_SPACE_NAMESPACE in namespace field? If it serializable a symbol cannot be used. Can we use a special string instead? What is the benefit of using a separate type?

Consumers of the Saved Objects Client can specify this for any operation they wish to perform explicitly against the default space. You bring up a good point about serializability that I hadn't considered -- @mikecote, do you foresee this being problematic for alerting? How are you planning/hoping to record the space/namespace which the alert should run in? Are you storing the space id, and translating to namespace when required, storing the namespace directly, or storing the base path and translating?

> DEFAULT_SPACE_ID exists here only for backward compatibility?

`DEFAULT_SPACE_ID` is mostly for backwards compatibility, but it also serves as a way to allow the spaces plugin to be completely disabled/re-enabled at any time, without breaking Kibana.

Reply:
> How are you planning/hoping to record the space/namespace which the alert should run in? Are you storing the space id, and translating to namespace when required, storing the namespace directly, or storing the base path and translating?

This will be a conversation I'll have to have with you when we do the security phase. Right now we store the `basePath` and create clients with that. Though there may be a few places that forget to store the space and assumes the default space.

Reply:
> This will be a conversation I'll have to have with you when we do the security phase. Right now we store the basePath and create clients with that. Though there may be a few places that forget to store the space and assumes the default space.

Ok, then this shouldn't be an issue for alerting then.

In general, specifying a custom namespace when using the saved objects client should be a very rare event. I consider the `namespace` field an implementation detail of the space plugin, even though it is an OSS concept. If the spaces plugin exposes first-class methods for translating a space id to a namespace (and vice-versa), then I don't foresee Symbols being problematic here.

> Can we use a special string instead? What is the benefit of using a separate type?

A special string would have to include some character that is [disallowed as part of a space id](https://github.com/elastic/kibana/blob/7ac8e4d9ccc2a9e064d8a5621cd62453d1006164/x-pack/legacy/plugins/spaces/server/lib/space_schema.ts#L11), in order to guarantee that we never have a collision. It's certainly doable, so long as we don't change the set of allowed characters in the future.

A symbol makes it painfully clear that we're dealing with a special case, as opposed to a mis-typed space id or similar.

Reply:
> I consider the namespace field an implementation detail of the space plugin, even though it is an OSS concept.

ok, agree. In this case symbol is internal detail and not a part of public contract. why it leaks to sibling plugins? https://github.com/elastic/kibana/pull/38444/files#diff-9b008eb2578ead2253f9faf64b262f0eR79


> If the spaces plugin exposes first-class methods for translating a space id to a namespace (and vice-versa), then I don't foresee Symbols being problematic here.

with this API can we have separate 2 types. don't we?
`type SavedObjectsNamespace = string | undefined;` in OSS
`type SavedObjectsNamespaceInternal = string | undefined | symbol;` in x-pack/spaces.

Reply:
> why it leaks to sibling plugins?

The type definition is declared in `x-pack/types`, which is shared by all x-pack plugins:https://github.com/elastic/kibana/blob/e2796ff47d64e7c994fb215e1a61e1004463cd85/x-pack/typings/core.d.ts#L8-L13

> with this API can we have separate 2 types. don't we?
type SavedObjectsNamespace = string | undefined; in OSS
type SavedObjectsNamespaceInternal = string | undefined | symbol; in x-pack/spaces.

I don't know if it's really internal though. The wrapped saved objects client has a public api that expects either `string | undefined | symbol`, and this is consumable by both first and third-party plugins.

Plus, having a different name for this vs the OSS version makes overriding the type very difficult, or maybe even impossible (I haven't tried yet). The current override (above) is rather straightforward.